Source/WebCore/css/typedom/numeric/CSSNumericType.cpp

```diff
@@ -0,0 +1,185 @@
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "config.h"
+#include "CSSNumericType.h"
+
+#if ENABLE(CSS_TYPED_OM)
+
+namespace WebCore {
+
+std::optional<CSSNumericType> CSSNumericType::addTypes(CSSNumericType a, CSSNumericType b)
+{
+    // https://drafts.css-houdini.org/css-typed-om/#cssnumericvalue-add-two-types
+    if (a.percentHint && b.percentHint && *a.percentHint != *b.percentHint)
+        return std::nullopt;
+
+    if (a.percentHint)
+        b.applyPercentHint(*a.percentHint);
+    if (b.percentHint)
+        a.applyPercentHint(*b.percentHint);
+
+    if (a == b)
+        return { WTFMove(a) };
+
+    for (auto type : eachBaseType()) {
+        if (type == CSSNumericBaseType::Percent)
+            continue;
+        if (!a.valueForType(type) && !b.valueForType(type))
+            continue;
+        a.applyPercentHint(type);
+        b.applyPercentHint(type);
+        if (a.valueForType(type) != b.valueForType(type))
+            return std::nullopt;
+    }
+
+    return { WTFMove(a) };
+}
+
+template<typename Argument> std::optional<CSSNumericType> typeFromVector(const Vector<Ref<CSSNumericValue>>& values, std::optional<CSSNumericType>(*function)(Argument, Argument))
+{
+    if (values.isEmpty())
+        return std::nullopt;
+    std::optional<CSSNumericType> type = values[0]->type();
+    for (size_t i = 1; i < values.size(); i++) {
+        type = function(*type, values[i]->type());
+        if (!type)
+            return std::nullopt;
+    }
+    return type;
+}
+
+std::optional<CSSNumericType> CSSNumericType::addTypes(const Vector<Ref<CSSNumericValue>>& values)
+{
+    return typeFromVector(values, addTypes);
+}
+
+std::optional<CSSNumericType> CSSNumericType::multiplyTypes(const CSSNumericType& a, const CSSNumericType& b)
+{
+    // https://drafts.css-houdini.org/css-typed-om/#cssnumericvalue-multiply-two-types
+    if (a.percentHint && b.percentHint && *a.percentHint != *b.percentHint)
+        return std::nullopt;
+
+    auto add = [] (auto left, auto right) -> CSSNumericType::BaseTypeStorage {
+        if (!left)
+            return right;
+        if (!right)
+            return left;
+        return *left + *right;
+    };
+    
+    return { {
+        add(a.length, b.length),
+        add(a.angle, b.angle),
+        add(a.time, b.time),
+        add(a.frequency, b.frequency),
+        add(a.resolution, b.resolution),
+        add(a.flex, b.flex),
+        add(a.percent, b.percent),
+        a.percentHint ? a.percentHint : b.percentHint
+    } };
+}
+
+std::optional<CSSNumericType> CSSNumericType::multiplyTypes(const Vector<Ref<CSSNumericValue>>& values)
+{
+    return typeFromVector(values, multiplyTypes);
+}
+
+String CSSNumericType::debugString() const
+{
+    return makeString("{",
```

Comment:
'{'

Reply:
I know this set be only used for debugging (based on the name) but it'd be really more efficient to use StringBuilder instead instead of makeString() with nested makeString() calls that create unnecessary temporary Strings.